### PR TITLE
Expose `org.jenkins.ui.symbol.Symbol`

### PR DESCRIPTION
**Pull request by Vlatombe**
In some cases, plugins need to generate markup with symbols (for example when generating `FormValidation` with html markup). This is adding a way for them to build such markup.

This also improves coverage for this method.

* Exposes the existing `IconSet#getSymbol` for plugins by adding a builder
pattern, moving it to its own package and class to separate the
Symbol framework from the Icon one.
* Added coverage to the existing method including symbol lookup from
  plugins

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

* Developer: allow plugins to generate symbol markup from Java code.
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6659"></a>

### CODE REVIEW

**Review by timja — APPROVED** [review]
Nice! Needed something like this in credentials plugin.

There's a fork of `IconSet` basically. For the menu's, only way I got it working was calling the restricted code but I left it out of the change set in the end

**Review by NotMyFault — APPROVED** [review]
Nice one

**Comment by daniel-beck** [comment]
To minimize problems integrating next week's core security fixes into jenkinsci/master, I'm on-holding this PR.

**Comment by daniel-beck** [comment]
Un-on-holding since the release is now out.

File: core/src/main/java/org/jenkins/ui/symbol/Symbol.java
Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java:66]: Also causes https://issues.jenkins.io/browse/JENKINS-68805 because the regexes make no sense.

Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java:66]: https://github.com/jenkinsci/jenkins/pull/6685 demonstrates the problem.

File: core/src/main/java/org/jenkins/ui/symbol/Symbol.java
Comment (Vlatombe) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java]: [f92ccaa](https://github.com/jenkinsci/jenkins/pull/6659/commits/f92ccaa3546896b6d09f48e9dc7b2b57ce2ef705) should address it

Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java]: Interesting, from looking at the code I'd expect it to restore the problematic behavior of caching caller-dependent attributes. I guess I'll see the test fail if I'm right :)

Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java]: I missed that, thanks! This should do it.

File: core/src/main/java/org/jenkins/ui/symbol/Symbol.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/symbol/Symbol.java:95]: @NotMyFault might this be the root cause of this PR instead? https://github.com/jenkinsci/jenkins/pull/6686

Comment (daniel-beck) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java:95]: `replace` replaces all, it's just that it's not a regex.

Comment (timja) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java:95]: > The dot of the price tag has no stroke defined, this replacement pattern wouldn't apply at all.

ah my bad.

it could be yes although I guess it's an assumption based on what ionicon's does

Comment (Vlatombe) [reply core/src/main/java/org/jenkins/ui/symbol/Symbol.java:95]: Looks like everything is fine so far. However having some logic to validate symbols at build time would be interesting.

What comes to my mind:
* Check that each `path` has at least one of `fill`,`stroke` defined and set to either `none` or `currentColor`

**Comment by Vlatombe** [comment]
Could I get a review from @jenkinsci/core-security-review ? This has been hanging for a while now.

File: test/src/test/java/org/jenkins/ui/symbol/SymbolJenkinsTest.java
Comment (basil) [reply test/src/test/java/org/jenkins/ui/symbol/SymbolJenkinsTest.java:17]: Not anymore: commit de5b030a629d1a489b9a010a224a53453b40f237

File: core/src/main/java/org/jenkins/ui/symbol/Symbol.java
Comment (daniel-beck) [inline core/src/main/java/org/jenkins/ui/symbol/Symbol.java:52]: Would it make more sense to move the compatibility code out of this new class and into callers that potentially pass legacy values? Then guaranteed modern callers wouldn't have this hack on the code path.

File: core/src/main/java/org/jenkins/ui/symbol/SymbolRequest.java
Comment (daniel-beck) [inline core/src/main/java/org/jenkins/ui/symbol/SymbolRequest.java:6]:
```suggestion
/**
 * @since TODO
 */
public final class SymbolRequest {
```

(Plus ideally more)

**Review by daniel-beck — COMMENTED** [review]
Is there a plugin in `jenkinsci` you could demonstrate use of this new API with?

Nit: the JUnit 5 comments should be resolved.

**Comment by Vlatombe** [comment]
> Is there a plugin in jenkinsci you could demonstrate use of this new API with?

Sorry I have other high priority tasks (and the initial use case that drove me to this was in a proprietary plugin)

**Comment by basil** [comment]
As far as I can tell the only pending task left on this PR is a validation of the proposed new API with a consumer. We have identified one possible consumer, the Credentials plugin. Given Vincent's limited availability, we would welcome any community contributions to adapt the Credentials plugin to the new API proposed in this PR.

**Comment by timja** [comment]
I've had a look at the credentials plugin and I'm not sure if this API is useful for it.

It's after an Icon not a symbol.

and fixing the issue there will likely require more refactoring in core around context menus and symbols.

I can't see a way with the current API to make it work right now.

**Comment by timja** [comment]
> Is that a use case that we would expect this API to work with

Unsure, the credentials plugin context menu code is not easy to reason with. I or someone else will need more time in there to try unravel it.

I think this API is a net positive, it would make that sort of thing easier to implement in the future.

**Comment by basil** [comment]
> Unsure

Fair enough. I don't think we should move forward with this PR until we become sure one way or another. Committing to an API that doesn't necessarily satisfy all use cases could be a liability in the future, since changing an API to accommodate new use cases while retaining compatibility once there are existing consumers is much harder (and in some cases impossible) compared to getting the design right at the outset.

File: core/src/test/java/org/jenkins/ui/symbol/SymbolTest.java
Comment (uhafner) [inline core/src/test/java/org/jenkins/ui/symbol/SymbolTest.java:27]: In order to see if that code would help for https://github.com/jenkinsci/prism-api-plugin/pull/59 I am wondering where you are actually checking for the SVG content? Can't you add an assertion for the most important part as well?

Comment (Vlatombe) [reply core/src/test/java/org/jenkins/ui/symbol/SymbolTest.java:30]: Should be okay now.

File: test/src/test/java/org/jenkins/ui/symbol/SymbolJenkinsTest.java
Comment (uhafner) [inline test/src/test/java/org/jenkins/ui/symbol/SymbolJenkinsTest.java:26]: This would not work out of the box in my use case. I have a given symbol as full referenced string with plugin included, e.g. `"symbol-solid/triangle-exclamation plugin-font-awesome-api"`. I can split this manually before and after the space but it would be simpler if another builder method would be available that takes the whole string. What do you think, does this make sense?

Comment (uhafner) [reply test/src/test/java/org/jenkins/ui/symbol/SymbolJenkinsTest.java:26]: I tried to use the new API in https://github.com/jenkinsci/prism-api-plugin/pull/60. However, it seems that I only get a cross as icon. It would be really helpful if we can add some logging why an icon is not found :-(

Or do I need to remove the `plugin-` suffix?

Comment (uhafner) [reply test/src/test/java/org/jenkins/ui/symbol/SymbolJenkinsTest.java:26]: Ok, I got it working now:

```
             String[] elements = StringUtils.split(name);
             if (elements.length == 2) {
                 String symbol = Symbol.get(new SymbolRequest.Builder()
                         .withName(StringUtils.removeStart(elements[0], "symbol-"))
                         .withPluginName(StringUtils.removeStart(elements[1], "plugin-"))
                         .withClasses(ICON_MD)
                         .build());
                 return new UnescapedText(symbol);
             }
```

From my perspective it would be helpful, if there would be another builder method (something like `withId` or `withIdentifier`), that takes the whole symbol identifier and splits it accordingly.

**Comment by uhafner** [comment]
@Vlatombe are you planning to finish this PR in the near future?

**Comment by uhafner** [comment]
> @uhafner I pushed a new method based on what you commented earlier. Let me know if it satisfies your use case.

Yes, this is exactly what I need. It seems that the incremental has not been published so I cannot test, but from the code and test case it looks good!

**Comment by uhafner** [comment]
I tested the new implementation, it works as expected! (See https://github.com/jenkinsci/prism-api-plugin/blob/599a3307af0e4e6263db95cee1ed79f7ae541131/src/main/java/io/jenkins/plugins/prism/SourcePrinter.java#L105-L113)

🎉

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
